lib/fuse_lowlevel.c

```diff
@@ -3009,6 +3039,36 @@ struct fuse_session *fuse_session_new(struct fuse_args *args,
 	return NULL;
 }
 
+int fuse_session_custom_io(struct fuse_session *se, const struct fuse_custom_io *io,
+			   int fd)
+{
+	if (fd < 0) {
+		fuse_log(FUSE_LOG_ERR, "Invalid file descriptor value %d passed to "
+			"fuse_session_custom_io()\n", fd);
+		return -EBADF;
+	}
+	if (io == NULL) {
+		fuse_log(FUSE_LOG_ERR, "No custom IO passed to "
+			"fuse_session_custom_io()\n");
+		return -EINVAL;
+	} else if (io->read == NULL || io->writev == NULL) {
```

Comment:
I'm worried about this invariant being safe. We're checking once when calling the function, but later changes to the code may not be aware of it, or someone may even modify the structure after the function has been called. So if we can find a way to avoid this, that'd be great.

Reply:
At the end of this function, the passed `fuse_custom_io` structure is actually copied (`*se->io = *io;`), so changes to the passed structure after this function is called, should have no impact.

Maybe `memcpy(se->io, io, sizeof(io))` would have been clearer instead of `*se->io = *io;`?